### PR TITLE
feat: Build three-way swap functionality (SC-447)

### PR DESCRIPTION
**Pull request by lucas-manuel**
Refactors to support three assets.

1. Update constructor
2. Remove dedicated swap/preview functions in favour of generic preview/swap functions
3. Remove single use `_getAssetsByValue` function and move logic directly into `convertToAssets`
4. Add `_isValidAsset` helper function to check if asset is one of the three assets
5. Remove single use `_divRoundUp` function and move logic into `_convertToSharesRoundUp`
6. Improve test coverage for conversion functions
7. Add tests for new preview/swap functions

### CODE REVIEW

File: test/Swaps.t.sol [hunk 1]
```diff
@@ -5,289 +5,327 @@ import "forge-std/Test.sol";
...
+    function test_swap_invalid_assetIn() public {
+        vm.expectRevert("PSM/invalid-asset");
+        psm.swap(makeAddr("other-token"), address(sDai), 100e6, 80e18, receiver);
+    }
+
+    function test_swap_invalid_assetOut() public {
+        vm.expectRevert("PSM/invalid-asset");
+        psm.swap(address(usdc), makeAddr("other-token"), 100e6, 80e18, receiver);
+    }
+
```
Comment (hexonaut) [inline test/Swaps.t.sol:42]: Should add a test to fail when the in and out assets are the same.

File: test/Swaps.t.sol [hunk 2]
```diff
@@ -5,289 +5,327 @@ import "forge-std/Test.sol";
...
-        assertEq(usdc.balanceOf(receiver),     0);
-        assertEq(usdc.balanceOf(address(psm)), 200e6);
+        assertEq(assetOut.balanceOf(receiver),     amountOut);
+        assertEq(assetOut.balanceOf(address(psm)), psmAssetOutBalance - amountOut);
     }
 
 }
 
-contract PSMSwapAssetOneToZeroTests is PSMTestBase {
+contract PSMSwapDaiAssetInTests is PSMSuccessTestsBase {
```
Comment (hexonaut) [inline test/Swaps.t.sol:158]: This contract is labelled as `PSMSwapDaiAssetInTests`, but also contains the USDC and sDAI tests. I'm assuming this is a mistake and you want `PSMSwapUSDCAssetInTests`, etc.

**Review by hexonaut — CHANGES_REQUESTED** [review]

File: src/PSM.sol [hunk 3]
```diff
@@ -19,59 +19,61 @@ contract PSM {
...
     uint256 public totalShares;
 
     mapping(address user => uint256 shares) public shares;
 
-    constructor(address asset0_, address asset1_, address rateProvider_) {
+    constructor(address asset0_, address asset1_, address asset2_, address rateProvider_) {
         require(asset0_       != address(0), "PSM/invalid-asset0");
         require(asset1_       != address(0), "PSM/invalid-asset1");
+        require(asset2_       != address(0), "PSM/invalid-asset2");
         require(rateProvider_ != address(0), "PSM/invalid-rateProvider");
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/PSM.sol:42]: Added

File: src/PSM.sol [hunk 4]
```diff
@@ -215,30 +233,59 @@ contract PSM {
...
+    }
+
+    function _previewSwapFromAsset2(uint256 amountIn, uint256 assetInPrecision)
+        internal view returns (uint256)
+    {
+        return amountIn
+            * IRateProviderLike(rateProvider).getConversionRate()
+            * assetInPrecision
+            / 1e27
+            / asset2Precision;
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/PSM.sol:276]: Updated function

File: test/Constructor.t.sol [hunk 5]
```diff
@@ -11,29 +11,36 @@ contract PSMConstructorTests is PSMTestBase {
...
+
+    function test_constructor_invalidAsset2() public {
+        vm.expectRevert("PSM/invalid-asset2");
+        new PSM(address(dai), address(usdc), address(0), address(rateProvider));
     }
 
     function test_constructor_invalidRateProvider() public {
         vm.expectRevert("PSM/invalid-rateProvider");
-        new PSM(address(sDai), address(usdc), address(0));
+        new PSM(address(dai), address(sDai), address(usdc), address(0));
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/Constructor.t.sol:29]: Updated

File: test/Getters.t.sol [hunk 6]
```diff
@@ -13,152 +13,157 @@ contract PSMHarnessTests is PSMTestBase {
...
+        assertEq(psmHarness.getAsset0Value(100e6), 100e6);
+        assertEq(psmHarness.getAsset0Value(200e6), 200e6);
+        assertEq(psmHarness.getAsset0Value(300e6), 300e6);
 
-        assertEq(psmHarness.getAsset0Value(100_000_000_000e6), 100_000_000_000e18);
-        assertEq(psmHarness.getAsset0Value(200_000_000_000e6), 200_000_000_000e18);
-        assertEq(psmHarness.getAsset0Value(300_000_000_000e6), 300_000_000_000e18);
+        assertEq(psmHarness.getAsset0Value(100_000_000_000e6), 100_000_000_000e6);
+        assertEq(psmHarness.getAsset0Value(200_000_000_000e6), 200_000_000_000e6);
+        assertEq(psmHarness.getAsset0Value(300_000_000_000e6), 300_000_000_000e6);
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/Getters.t.sol:35]: Yep sorry fixed

File: src/PSM.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/PSM.sol:42]: Added test coverage as well

File: test/Previews.t.sol [hunk 7]
```diff
@@ -5,98 +5,129 @@ import "forge-std/Test.sol";
...
+        psm.previewSwap(makeAddr("other-token"), address(usdc), 1);
+    }
 
-        assertEq(psm.previewSwapAssetZeroToOne(1e6), 0.8e18);
-        assertEq(psm.previewSwapAssetZeroToOne(2e6), 1.6e18);
-        assertEq(psm.previewSwapAssetZeroToOne(3e6), 2.4e18);
+    function test_previewSwap_invalidAssetOut() public {
+        vm.expectRevert("PSM/invalid-asset");
+        psm.previewSwap(address(usdc), makeAddr("other-token"), 1);
+    }
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/Previews.t.sol:19]: Added

File: test/Swaps.t.sol (hunk 1 above)
Comment (lucas-manuel) [reply test/Swaps.t.sol:42]: Added for all three

File: test/Previews.t.sol [hunk 8]
```diff
@@ -78,15 +90,13 @@ contract PSMPreviewSwapUSDCAssetInTests is PSMTestBase {
...
     }
 
-    function testFuzz_previewSwap_daiToSDai(uint256 amountIn, uint256 conversionRate) public {
-        amountIn = _bound(amountIn, 0, USDC_TOKEN_MAX);
-
-        amountIn       = _bound(amountIn,       1,       10_000_000_000e18);  // Using 10 billion for conversion rates
-        conversionRate = _bound(conversionRate, 0.01e27, 1000e27);            // 1% to 100,000% conversion rate
+    function testFuzz_previewSwap_usdcToSDai(uint256 amountIn, uint256 conversionRate) public {
+        amountIn       = _bound(amountIn,      1,         USDC_TOKEN_MAX);
+        conversionRate = bound(conversionRate, 0.0001e27, 1000e27);  // 0.01% to 100,000% conversion rate
```
Comment (hexonaut) [inline test/Previews.t.sol:95]: This should be `_bound` for consistency.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Got a fuzz revert:

```
[FAIL. Reason: revert: PSM/amountOut-too-low; counterexample: calldata=0x4e362699000000000000000000000000000000000000000000000000000001d9a4d0603b0000000000000000000000000000000000000000000000000000000000000055000000000000000000000000199bd4c0e7a7a4e756b74ca70577a9d232375f310000000000000000000000006b9f40c73f758c3bb06da738460f8c8d85a857b7 args=[2034284650555 [2.034e12], 85, 0x199BD4c0E7A7A4e756B74Ca70577a9D232375F31, 0x6B9f40c73F758c3bB06Da738460f8c8d85A857b7]] testFuzz_swap_usdcToSDai(uint256,uint256,address,address) (runs: 0, μ: 0, ~: 0)
```

**Review by hexonaut — APPROVED** [review]

**Pull request merged**